core/threaded/scheduler_GEDF_NP.c

```diff
@@ -223,91 +105,138 @@ void lf_sched_init(environment_t* env, size_t number_of_workers, sched_params_t*
+                       LF_LEVEL(reaction_to_return->index), scheduler->custom_data->current_level);
+        // We need to wait to advance to the next level or get a new reaction at the current level.
+        if (scheduler->number_of_idle_workers == scheduler->number_of_workers - 1) {
+          // All other workers are idle.  Advance to the next level.
+          if (++scheduler->custom_data->current_level > scheduler->max_reaction_level) {
+            // Since the reaction queue is not empty, we must be cycling back to level 0 due to deadlines
+            // having been given precedence over levels.  Reset the current level to 1.
+            scheduler->custom_data->current_level = 0;
+          }
+          LF_PRINT_DEBUG("Scheduler: Advancing to next reaction level %zu.", scheduler->custom_data->current_level);
+#ifdef FEDERATED
+          // In case there are blocking network input reactions at this level, stall.
+          lf_stall_advance_level_federation_locked(scheduler->custom_data->current_level);
+#endif
+        } else {
+          // Some workers are still working on reactions on the current level.
+          // Wait for them to finish.
+          wait_for_other_workers_to_finish(scheduler, worker_number);
+        }
+      }
+    } else {
+      // The reaction queue is empty.
+      LF_PRINT_DEBUG("Worker %d finds nothing on the reaction queue.", worker_number);
+
+      // If all other workers are idle, then we are done with this tag.
+      if (scheduler->number_of_idle_workers == scheduler->number_of_workers - 1) {
+        // Last thread to go idle
+        LF_PRINT_DEBUG("Scheduler: Worker %d is advancing the tag.", worker_number);
+        // Advance the tag.
+        // Set a flag in the scheduler that the lock is held by the sole executing thread.
+        // This prevents acquiring the mutex in lf_scheduler_trigger_reaction.
+        scheduler->custom_data->solo_holds_mutex = true;
+        if (_lf_sched_advance_tag_locked(scheduler)) {
+          LF_PRINT_DEBUG("Scheduler: Reached stop tag.");
+          scheduler->should_stop = true;
+          scheduler->custom_data->solo_holds_mutex = false;
+          // Notify all threads that the stop tag has been reached.
+          LF_COND_BROADCAST(&scheduler->custom_data->reaction_q_changed);
+          break;
+        }
+        scheduler->custom_data->solo_holds_mutex = false;
+        // Reset the level to 0.
+        scheduler->custom_data->current_level = 0;
+#ifdef FEDERATED
+        // In case there are blocking network input reactions at this level, stall.
+        lf_stall_advance_level_federation_locked(scheduler->custom_data->current_level);
+#endif
+      } else {
+        // Some other workers are still working on reactions on the current level.
+        // Wait for them to finish.
+        wait_for_other_workers_to_finish(scheduler, worker_number);
+      }
```

Comment:
This could be factored out into a separate `static` procedure with a signature like `advance_tag_or_wait(lf_scheduler_t* scheduler, int worker_number)`